Vue3 Migration from Vue 2

gerteck opened the pull request:
**What is the purpose of this pull request?**

- [x] Documentation update
- Bug fix
- [x] Feature addition or enhancement
- [x] Code maintenance
- DevOps
- Improve developer experience
- Others, please explain:

<!--
  If this pull request is addressing an issue, link to the issue: #xxx

  If this pull request completely addresses an issue, use one of the closing keywords: "Fixes #xxx" or "Resolves #xxx"
  https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword

  Otherwise, elaborate further on the rationale of this pull request as needed
-->

**Related PRs for Reference**:
https://github.com/MarkBind/markbind/pull/1534

**Overview of changes:**

MarkBind (v5.5.3) is currently using Vue 2. However, Vue 2 has reached EOL and limits the extensibility and maintainability of MarkBind, especially the vue-components package. (UI Library Package). 

This PR tracks the Vue Migration Vue 2 to Vue 3.

Disgustingly intricate understanding of both Vue 2, Vue 3, low-level dependency APIs and MarkBind (in particular with regards to SSR, Hydrating, Bundling) is needed to conduct this.   

Some relevant links to understanding / reviewing the PR:
* [Vue 3 API Reference](https://vuejs.org/api/application.html#createapp)
* [Vue SFC](https://vuejs.org/guide/scaling-up/sfc.html)
* [Vue Rendering Mechanism](https://vuejs.org/guide/extras/rendering-mechanism)
* [Vue Migration Build](https://v3-migration.vuejs.org/migration-build)
* [@vue/compiler-sfc](https://www.npmjs.com/package/@vue/compiler-sfc)
* [@vue/server-renderer](https://www.npmjs.com/package/@vue/server-renderer)

Additionally, here are some relevant documentation of the codebase I have written and compiled to speed up the understanding process for purpose of peer development/review.
* [pageVueServerRenderer](https://gerteck.github.io/mb-architecture/contents/core/pageVueServerRenderer.html)
* [Core-Web Package](https://gerteck.github.io/mb-architecture/contents/core-web/core-web.html)
* [Vue-Components](https://gerteck.github.io/mb-architecture/contents/vue/vueComponents.html)

Note that currently, MarkBind Vue components are authored in the Vue Options API style. While migrating to Vue 3, we can continue to use this API style. Future work can consider cost benefit of switching to using Composition API.

will update desc further when i compile the information needed

left to do: 
* [ ] refactoring of vue components (continue using options API), 
* [ ] refactoring of plugins, wrt. Vue directives (breaking changes).
* [ ] extensive update of tests afterwards
* [ ] extensive documentation updates if and as needed

Fixes #2084
Fixes #2597
Fixes #2171 (Took the chance to fix w. changes in vue rendering APIs) 

tbc..

**Anything you'd like to highlight/discuss:**



**Testing instructions:**

<!--
  Any special testing instructions **not including** our automated tests.
-->

**Proposed commit message: (wrap lines at 72 characters)**

Migrate Vue 2 to Vue 3

tbc

<!--
  See this link for more info on how to write a good commit message:
  https://se-education.org/guides/conventions/git.html

  As best as possible, write a succinct commit title in 50 characters

|---------This is the width of 50 chars----------|
|-----------This is the width of 72 chars for your reference-----------|
-->

---

**Checklist:** :ballot_box_with_check:

<!-- Leave non-applicable items unchecked -->

- [ ] Updated the documentation for feature additions and enhancements
- [ ] Added tests for bug fixes or features
- [ ] Linked all related issues
- [ ] No unrelated changes <!-- It's tempting, but increases the reviewer's work, and really pollutes the commit history =( -->

<!--
  We'll try our best to get to your PR within a week.
  If we haven't gotten to it then, or if your pull request resolves an urgent item, feel free to give us a ping!
-->

---

**Reviewer checklist:**

Indicate the [SEMVER](https://semver.org/) impact of the PR:
- [ ] Major (when you make incompatible API changes)
- [ ] Minor (when you add functionality in a backward compatible manner)
- [ ] Patch (when you make backward compatible bug fixes)

At the end of the review, please label the PR with the appropriate label: `r.Major`, `r.Minor`, `r.Patch`.

Breaking change release note preparation (if applicable):
- To be included in the release note for any feature that is made obsolete/breaking

> Give a brief explanation note about:
>   - what was the old feature that was made obsolete
>   - any replacement feature (if any), and
>   - how the author should modify his website to migrate from the old feature to the replacement feature (if possible).

gerteck commented:
https://vuejs.org/guide/extras/rendering-mechanism#templates-vs-render-functions

Opted to use templates instead of render functions as templates are simpler, and still conduct SSR.

tlylt commented:
Hi @gerteck, thanks for taking this up! As you have probably read up on all the related past issues/PRs around the migration, I would just want to caution that please take into account of how other developers/reviewers can help in this migration process. If the work is not broken down into manageable chunks, it might be difficult review and assess the changes holistically.

gerteck commented:
I am currently facing issues which @EltonGohJH previously faced:

> `resolveComponent can only be used in render() or setup() error.`
> Occurs due to multiple Vue instances.? For SSR, the issue persists because two different Vue files are used (runtime-core.cjs.js and runtime-core.esm-bundler.js). This makes it non-trivial to solve, so SSR has been disabled for now. This is not trivial to debug.

Not too sure how to resolve this issue, but am looking into the matter.

Secondly, <overlay> and related components that utilize portal vue also face the issue of `Cannot read property 'isCE' of null in remote component with slot `. Preliminary [research](https://github.com/vuejs/core/issues/4344) suggests that this issue is also a result of having multiple Vue Instances.

lhw-1 commented:
Just to add on a small comment, the first issue is caused [here](https://github.com/vuejs/core/blob/main/packages/runtime-core/__tests__/helpers/resolveAssets.spec.ts#L99). Certainly not trivial to debug, and we'll need to look into if it is worth spending more time on the issue this semester.

gerteck commented:
I have reformatted and opened a new PR for the Vue migration #2622 . It has been structured to be as easy to follow as possible. 

As advised by @tlylt , I have made best effort to break the work done into manageable chunks for reviewers and other developers.

I have isolated separable code changes into separate PRs as much as possible, which have since been merged. 
* (PRs: 2625, 2626, 2614). 
* The final PR2622 contains the minimal (Best Effort) amount of code changes to functionally migrate Vue 2 to Vue 3.

Additionally, in the PR and the description, currently, I have:
* Modularized the commits - each commit deals with a logically separable component in the migration process. 
* Added an overview and detailed description for each commit - what is done with each commit, grouped logically. 
* Hyperlinked where convenient — Direct links to referenced commits, packages, external documentation to help in navigating quickly and access necessary context.
* Added release note preparation, some suggested testing instruction, related issues for context at appropriate places and some other stuff.

Hope it makes it easier on PR reviewers! Will close this PR.
